MigrationTutorial/Blog-Post.md

```diff
@@ -0,0 +1,658 @@
+
+#### Migration function
+A new detail emerges in this migration function: checking against `oldSchemaVersion` *21)*. This is necessary because a migration must be able to be executed also when done from version X to version X+2 or higher. In our case from version 1 straight to 3. The content of the `if` branch *21)* is exactly the content of the migration from V1 to V2.  
+After that, we take care of reclassifying `GlueHolder` *22)* and `Brush` *23)*. In order to do this we need to obtain the old `Consumable` of the needed type *24)*, extract all the fields that are relevant to `MachineryAndTool` *25)* and create a new instance of the latter class to be filled with the extracted data, where appropriate *26)*. When all of this is done, the new instance of `MachineryAndTool` is added to the new realm *26)* while the previous instance of `Consumable` is removed from the realm *27)*, given that its logical representation was moved to another class (`MachineryAndTool`).
+
+</br>
+
+---
+#### *(14) V3 Migration function*
+```cs
+public static void DoMigrate(Migration migration, ulong oldSchemaVersion)
+{
+    if (oldSchemaVersion < 2)   // 21)
+    {
+        var newEmployees = migration.NewRealm.All<Employee>();
+        var oldEmployees = migration.OldRealm.DynamicApi.All("Employee");
+
+        for (var i = 0; i < newEmployees.Count(); i++)
+        {
+            var newEmployee = newEmployees.ElementAt(i);
+            var oldEmployee = oldEmployees.ElementAt(i);
+            if (string.Equals(oldEmployee.Gender, "female", StringComparison.OrdinalIgnoreCase))
+            {
+                newEmployee.Gender = Gender.Female;
+            }
+            else if (string.Equals(oldEmployee.Gender, "male", StringComparison.OrdinalIgnoreCase))
+            {
+                newEmployee.Gender = Gender.Male;
+            }
+            else
+            {
+                newEmployee.Gender = Gender.Other;
+            }
+        }
+
+        var newConsumables = migration.NewRealm.All<Consumable>();
+        var oldConsumables = migration.OldRealm.DynamicApi.All("Consumable");
+        var distinctConsumableId = new HashSet<string>();
+        var consumableToDelete = new List<Consumable>();
+
+        migration.RenameProperty(nameof(Consumable), "Price", nameof(Consumable.LastPurchasedPrice));
+
+        for (var i = 0; i < newConsumables.Count(); i++)
+        {
+            var currConsumable = newConsumables.ElementAt(i);
+
+            // remove duplicates since ProductId is the new PrimaryKey
+            if (distinctConsumableId.Contains(currConsumable.ProductId))
+            {
+                consumableToDelete.Add(currConsumable);
+            }
+            else
+            {
+                distinctConsumableId.Add(currConsumable.ProductId);
+            }
+        }
+
+        consumableToDelete.ForEach(x => migration.NewRealm.Remove(x));
+    }
+
+    ConvertConsumableToTool(migration, oldSchemaVersion, "GlueHolder"); // 22)
```

Comment:
Are we assuming here that is at most one consumable of type "GlueHolder" (and "Brush")?

Reply:
Yes, each consumable has an `Amount`.